packages/flutter/lib/src/material/notches.dart

```diff
@@ -0,0 +1,174 @@
+import 'dart:math' as math;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+
+/// Computes notches in the outline of a shape.
+///
+/// Typically used to compute notches by a [BottomAppBar].
+abstract class Notch {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const Notch();
+
+  /// Returns a path for a notch in the outline of a shape.
+  ///
+  /// The path makes a notch in the host shape that can contain the guest shape.
+  ///
+  /// The `host` is the bounding rectangle for the shape into which the notch will
+  /// be applied. The `guest` is the bounding rectangle of the shape for which we
+  /// are creating a notch in the host.
+  ///
+  /// The `start` and `end` arguments are points on the outline of the host shape
+  /// that will be connected by the returned path.
+  ///
+  /// The returned path may pass anywhere, including inside the guest bounds area,
+  /// and may contain multiple subpaths. The returned path ends at `end` and does
+  /// not end with a [Path.close]. The returned [Path] is built under the
+  /// assumption it will be added to an existing path that is at the `start`
+  /// coordinates using [Path.addPath].
+  Path getPath(Rect host, Rect guest, Offset start, Offset end);
+}
+
+/// Computes smooth circular notches.
+///
+/// The notches computed by this class are paths that smoothly connect a top
+/// horizontal edge with a circular arc.
+///
+/// The notch computed by this class is intended to contain a circle bound by a
+/// square of the guest + [notchMargin] size, and will not be smaller than this
+/// circle.
+///
+/// The guest shape is assumed to be a circle (hence the guest rectangle is
+/// assumed to be a square).
+///
+/// See also: [getPath].
+class CircularNotch implements Notch {
+  /// Constructs a [CircularNotch] that can be used for multiple notch
+  /// computations.
+  const CircularNotch({
+    this.notchMargin = 4.0,
+  });
+
+  /// Minimal margin that will be kept from the guest. Assuming that the guest's
```

Comment:
The margin between the circular guest and the host's circular notch.

Reply:
You suggestion is easier to parse.
Though if we want to be accurate we need to say it's the minimal margin (as around the rounded corners the margin might be bigger).

I'm fine with both, let me know which one you prefer.

Reply:
I think is OK to be a little less precise here. You could say that we're computing a [Minkowski sum](http://hansmuller-webkit.blogspot.com/2013/04/growing-and-shrinking-polygons-round-one.html) but that wouldn't really help anyone :-)

Reply:
Wow :smiley:

Reply:
done